File: python_part/Leetcode/Data Structure/BinaryTree/easy/104. Maximum Depth of Binary Tree/Solution1.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None



'''
最大深度问题是个递归问题，可以使用DFS的思想来解决问题：

递归公式：

    maxDepth(T) = 1+max(maxDepth(T.left), maxDepth(T.right))
    
根据以上递归公式，可以得到递归出口为：

    if T is None -->  return 0
    
    
'''

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0

        return max(self.maxDepth(root.left), self.maxDepth(root.right))+1

File: Cpp_part/Leetcode/Leetcode/tree /94. 二叉树的中序遍历/solution1.cpp
#include <vector>
#include <iostream>

using namespace std;


 // Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;       //这是ptr，在python里面的绝大部分obj在c里面都是ptr
    TreeNode *right;      //简单理解就是 带有* 的后面的变量就是以ptr形式存放的obj。那么使用的时候就用 -> 就行了。
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> path;
        this -> helper(root, path);
        return path;
    }
    void helper(const TreeNode* root, vector<int>& path){
        if (root == NULL) return;
        this -> helper(root -> left, path);
        path.push_back(root -> val);
        this -> helper(root -> right,path);
    }
    
};
